File: solutions/typescript/perfect-numbers/1/perfect-numbers.ts
type Nicomachus = "perfect" | "abundant" | "deficient";

export function classify(n: number): Nicomachus {
  if (n <= 0) {
    throw new Error("Classification is only possible for natural numbers.");
  }

  const sum = aliquotSum(n);

  if (n == sum) return "perfect";
  return n < sum ? "abundant" : "deficient";
}

function aliquotSum(n: number): number {
  if (n == 1) return 0;

  let sum = 1;
  const sqrt = Math.sqrt(n);

  for (let i = 2; i <= sqrt; i++) {
    if (n % i == 0) {
      sum += i;
      const quotient = n / i;
      if (quotient != i) sum += quotient;
    }
  }

  return sum;
}

File: solutions/elixir/perfect-numbers/1/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number <= 0 do
    {:error, "Classification is only possible for natural numbers."}
  end

  def classify(1), do: {:ok, :deficient}
  def classify(2), do: {:ok, :deficient}

  def classify(number) do
    number
    |> factors()
    |> Enum.sum()
    |> then(fn
      sum when sum > number -> {:ok, :abundant}
      sum when sum < number -> {:ok, :deficient}
      _ -> {:ok, :perfect}
    end)
  end

  defp factors(number) do
    2..(number - 1)
    |> Enum.reduce([1], fn
      n, acc when rem(number, n) == 0 -> [n | acc]
      _, acc -> acc
    end)
  end
end
